File: frontend/js/views/workers_comp_premium.js
// Workers' comp premium generator — per class-code manual premium (payroll/100 ×
// rate), experience modifier, via /calc/workers-comp-premium.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

const SEED = [
    { code: '8810', description: 'Clerical', payroll: 300000, rate: 2.00 },
    { code: '5403', description: 'Carpentry', payroll: 200000, rate: 5.00 },
];

function rowHtml(c) {
    return `
        <div class="mpb-row wcp-row">
            <input type="text" class="wcp-code" placeholder="${esc(t('view.wcp.ph.code'))}" value="${esc(c.code || '')}">
            <input type="text" class="wcp-desc" placeholder="${esc(t('view.wcp.ph.desc'))}" value="${esc(c.description || '')}">
            <input type="number" step="1000" min="0" class="wcp-payroll" value="${c.payroll}">
            <input type="number" step="0.01" min="0" class="wcp-rate" value="${c.rate}">
            <button type="button" class="wcp-del" data-i18n="view.wcp.remove">Remove</button>
        </div>`;
}

export async function renderWorkersComp(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.wcp.h1.title">// WORKERS' COMP PREMIUM</span></h1>
        <p class="muted small" data-i18n="view.wcp.hint.intro">
            The standard premium estimate for a workers' compensation policy. Each class code carries a rate
            per $100 of payroll; the manual premium is payroll ÷ 100 × rate, summed across class codes, and the
            modified premium applies the employer's experience modifier (below 1.0 is a credit, above 1.0 a
            surcharge). Drafting aid, not insurance/legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.wcp.h2.inputs">Policy details</h2>
            <form id="wcp-form" class="inline-form">
                <label><span data-i18n="view.wcp.label.insurer">Insurer</span>
                    <input type="text" name="insurer_name" value="State Fund" required></label>
                <label><span data-i18n="view.wcp.label.employer">Employer</span>
                    <input type="text" name="employer_name" value="Acme Co" required></label>
                <label><span data-i18n="view.wcp.label.period">Policy period</span>
                    <input type="text" name="policy_period" value="2026" required></label>
                <label><span data-i18n="view.wcp.label.emod">Experience modifier</span>
                    <input type="number" step="0.01" min="0" name="experience_mod" value="0.90"></label>
                <label><span data-i18n="view.wcp.label.date">Date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.wcp.label.note">Note (optional)</span>
                    <input type="text" name="note" value="" placeholder="${esc(t('view.wcp.ph.note'))}"></label>
            </form>
            <div class="mpb-head wcp-head">
                <span data-i18n="view.wcp.col.code">Class</span>
                <span data-i18n="view.wcp.col.desc">Description</span>
                <span data-i18n="view.wcp.col.payroll">Payroll ($)</span>
                <span data-i18n="view.wcp.col.rate">Rate/100</span>
                <span></span>
            </div>
            <div id="wcp-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="wcp-add" class="secondary" data-i18n="view.wcp.add">+ Add class code</button>
        </div>
        <div id="wcp-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#wcp-form');
    const rowsEl = mount.querySelector('#wcp-rows');

    const generate = async () => {
        const fd = new FormData(form);
        const classes = [...rowsEl.querySelectorAll('.wcp-row')].map((r) => ({
            code: (r.querySelector('.wcp-code').value || '').trim(),
            description: (r.querySelector('.wcp-desc').value || '').trim(),
            payroll_usd: Number(r.querySelector('.wcp-payroll').value) || 0,
            rate_per_100_usd: Number(r.querySelector('.wcp-rate').value) || 0,
        })).filter((c) => c.code);
        const body = {
            insurer_name: (fd.get('insurer_name') || '').trim(),
            employer_name: (fd.get('employer_name') || '').trim(),
            policy_period: (fd.get('policy_period') || '').trim(),
            classes,
            experience_mod: Number(fd.get('experience_mod')) || 0,
            date: fd.get('date'),
            note: (fd.get('note') || '').trim(),
        };
        try {
            const doc = await api.calcWorkersComp(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.wcp.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);

    mount.querySelector('#wcp-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ code: '', description: '', payroll: 0, rate: 0 }));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('wcp-del')) {
            e.target.closest('.wcp-row').remove();
            generate();
        }
    });
    form.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('input', () => { live(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const rows = doc.rows.map((r) => `
        <tr><td>${esc(r.code)}</td><td>${esc(r.description)}</td><td>${money(r.payroll_usd)}</td>
            <td>${r.rate_per_100_usd}</td><td>${money(r.manual_premium_usd)}</td></tr>
    `).join('');
    const el = mount.querySelector('#wcp-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.wcp.card.modified">Modified premium</div>
                    <div class="value">${money(doc.modified_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.wcp.card.manual">Manual premium</div>
                    <div class="value">${money(doc.manual_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.wcp.card.emod">Exp. modifier</div>
                    <div class="value">${doc.experience_mod}</div></div>
                <div class="card"><div class="label" data-i18n="view.wcp.card.payroll">Total payroll</div>
                    <div class="value">${money(doc.total_payroll_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="wcp-copy" type="button" data-i18n="view.wcp.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="wcp-download" type="button" data-i18n="view.wcp.btn.download">Download .txt</button>
            </div>
        </div>
        <table class="data-table">
            <thead><tr>
                <th data-i18n="view.wcp.th.code">Class</th>
                <th data-i18n="view.wcp.th.desc">Description</th>
                <th data-i18n="view.wcp.th.payroll">Payroll</th>
                <th data-i18n="view.wcp.th.rate">Rate/100</th>
                <th data-i18n="view.wcp.th.premium">Premium</th>
            </tr></thead>
            <tbody>${rows}</tbody>
        </table>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#wcp-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.wcp.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.wcp.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#wcp-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'workers-comp-premium.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/workers_comp_premium.rs
//! Workers' compensation premium — the standard premium estimate for a business's
//! workers' comp policy. Each class code carries a rate per $100 of payroll; the
//! manual premium is payroll ÷ 100 × rate, summed across class codes, and the
//! modified premium applies the employer's experience modifier (a mod below 1.0 is
//! a credit for good loss history; above 1.0 a surcharge). No existing generator
//! computes a payroll-based class-rate premium. Drafting aid, not insurance/legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ClassCode {
    pub code: String,
    #[serde(default)]
    pub description: String,
    /// Annual payroll assigned to this class code.
    pub payroll_usd: f64,
    /// Rate per $100 of payroll for this class code.
    pub rate_per_100_usd: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct WorkersCompInput {
    pub insurer_name: String,
    pub employer_name: String,
    pub policy_period: String,
    pub classes: Vec<ClassCode>,
    /// Experience modifier (1.0 = neutral; <1 credit; >1 surcharge).
    #[serde(default = "default_emod")]
    pub experience_mod: f64,
    pub date: String,
    #[serde(default)]
    pub note: String,
}

fn default_emod() -> f64 {
    1.0
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ClassRow {
    pub code: String,
    pub description: String,
    pub payroll_usd: f64,
    pub rate_per_100_usd: f64,
    /// payroll ÷ 100 × rate.
    pub manual_premium_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct WorkersCompPremium {
    pub title: String,
    pub total_payroll_usd: f64,
    /// Sum of each class code's manual premium.
    pub manual_premium_usd: f64,
    pub experience_mod: f64,
    /// Manual premium × experience modifier.
    pub modified_premium_usd: f64,
    pub rows: Vec<ClassRow>,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &WorkersCompInput) -> WorkersCompPremium {
    let rows: Vec<ClassRow> = i
        .classes
        .iter()
        .map(|c| ClassRow {
            code: c.code.clone(),
            description: c.description.clone(),
            payroll_usd: cents(c.payroll_usd),
            rate_per_100_usd: c.rate_per_100_usd,
            manual_premium_usd: cents(c.payroll_usd / 100.0 * c.rate_per_100_usd),
        })
        .collect();

    let total_payroll = cents(i.classes.iter().map(|c| c.payroll_usd).sum());
    let manual = cents(rows.iter().map(|r| r.manual_premium_usd).sum());
    let modified = cents(manual * i.experience_mod);

    let detail = if rows.is_empty() {
        "No class codes listed.".to_string()
    } else {
        rows.iter()
            .map(|r| {
                let desc = if r.description.trim().is_empty() {
                    String::new()
                } else {
                    format!(" ({})", r.description)
                };
                format!(
                    "  • Class {}{}: payroll {} × {:.4}/100 = {}",
                    r.code, desc, money(r.payroll_usd), r.rate_per_100_usd, money(r.manual_premium_usd)
                )
            })
            .collect::<Vec<_>>()
            .join("\n")
    };

    let mod_desc = if i.experience_mod < 1.0 {
        "a credit"
    } else if i.experience_mod > 1.0 {
        "a surcharge"
    } else {
        "neutral"
    };

    let mut clauses = vec![
        DocClause {
            heading: "Policy".into(),
            body: format!(
                "Insurer: {}\nEmployer: {}\nPolicy period: {}\nDate: {}",
                i.insurer_name, i.employer_name, i.policy_period, i.date
            ),
        },
        DocClause { heading: "Class Codes".into(), body: detail },
        DocClause {
            heading: "Manual Premium".into(),
            body: format!(
                "Total payroll {} produces a manual premium of {} across the class codes above.",
                money(total_payroll), money(manual)
            ),
        },
        DocClause {
            heading: "Experience Modification".into(),
            body: format!(
                "The experience modifier of {:.3} ({}) applies, for a modified premium of {}.",
                i.experience_mod, mod_desc, money(modified)
            ),
        },
    ];

    let note = i.note.trim();
    if !note.is_empty() {
        clauses.push(DocClause { heading: "Note".into(), body: note.to_string() });
    }

    WorkersCompPremium {
        title: "Workers' Compensation Premium Estimate".into(),
        total_payroll_usd: total_payroll,
        manual_premium_usd: manual,
        experience_mod: i.experience_mod,
        modified_premium_usd: modified,
        rows,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn cls(code: &str, payroll: f64, rate: f64) -> ClassCode {
        ClassCode { code: code.into(), description: String::new(), payroll_usd: payroll, rate_per_100_usd: rate }
    }

    fn base() -> WorkersCompInput {
        WorkersCompInput {
            insurer_name: "State Fund".into(),
            employer_name: "Acme Co".into(),
            policy_period: "2026".into(),
            classes: vec![cls("8810", 300_000.0, 2.00), cls("5403", 200_000.0, 5.00)],
            experience_mod: 0.90,
            date: "2026-07-01".into(),
            note: String::new(),
        }
    }

    #[test]
    fn multi_class_manual_and_modified() {
        let d = generate(&base());
        assert!(close(d.total_payroll_usd, 500_000.0));
        // 300k/100×2 = 6,000; 200k/100×5 = 10,000 → 16,000.
        assert!(close(d.manual_premium_usd, 16_000.0));
        // 16,000 × 0.90 = 14,400.
        assert!(close(d.modified_premium_usd, 14_400.0));
    }

    #[test]
    fn neutral_mod_equals_manual() {
        let d = generate(&WorkersCompInput {
            classes: vec![cls("0000", 500_000.0, 3.50)],
            experience_mod: 1.0,
            ..base()
        });
        assert!(close(d.manual_premium_usd, 17_500.0));
        assert!(close(d.modified_premium_usd, 17_500.0));
    }

    #[test]
    fn surcharge_mod_increases_premium() {
        let d = generate(&WorkersCompInput {
            classes: vec![cls("0000", 500_000.0, 3.50)],
            experience_mod: 1.25,
            ..base()
        });
        assert!(close(d.modified_premium_usd, 21_875.0));
    }

    #[test]
    fn per_class_premium_rows() {
        let d = generate(&base());
        assert_eq!(d.rows.len(), 2);
        assert!(close(d.rows[0].manual_premium_usd, 6_000.0));
        assert!(close(d.rows[1].manual_premium_usd, 10_000.0));
    }

    #[test]
    fn empty_classes_zero() {
        let d = generate(&WorkersCompInput { classes: vec![], ..base() });
        assert!(close(d.manual_premium_usd, 0.0));
        assert!(d.clauses.iter().any(|c| c.body.contains("No class codes")));
    }

    #[test]
    fn note_appended_when_present() {
        let d = generate(&WorkersCompInput { note: "Audit at period end.".into(), ..base() });
        assert!(d.clauses.iter().any(|c| c.heading == "Note" && c.body.contains("Audit")));
    }
}
